### ISSUE TITLE
Add Docstrings for All Classes Across BardGPT Codebase

### ISSUE THREAD

**Issue by Himanshu7921** [Himanshu7921/BardGPT#2]
## 1. Issue Type

* [ ] Bug Report
* [x] Feature Request
* [ ] Documentation Improvement
* [ ] Training / Sampling Issue
* [ ] Performance / Optimization Request
* [ ] Other

---

## 2. Description

The BardGPT codebase currently lacks **consistent docstrings for many classes**, including core components like:

* Model classes (`DecoderOnlyTransformerModel`, attention blocks, feed-forward blocks, embeddings, etc.)
* Data loader and utility classes
* Training-related classes or helper objects
* Checkpoint / scheduler utilities

This makes it harder for contributors and new users to understand class responsibilities, inputs/outputs, and architectural decisions.

This issue proposes adding **proper, standardized docstrings** for every class in the repository.

---

## 3. Steps to Reproduce (for bugs)

*Not a bug — documentation enhancement request.*

---

## 4. Environment Information

* OS: Any
* Python version: N/A
* PyTorch version: N/A
* CUDA version: N/A
* GPU model: N/A
* BardGPT commit hash or version: Latest
* Installed dependencies: N/A

---

## 5. Additional Context

Classes currently missing docstrings:

* `DecoderOnlyTransformerModel`
* `MultiHeadAttention`
* `FeedForward`
* `TransformerBlock`
* `PositionalEmbedding`
* `TokenEmbedding`
* `DataLoader`
* `TransformerConfig`
* `generate()` utilities
* Any other helper classes/modules inside `utils/` and `model/`

**Expected Docstring Format (example)**:

```python
class DecoderOnlyTransformerModel(nn.Module):
    """
    A GPT-style decoder-only Transformer.

    Args:
        max_seq_length (int): Maximum context length.
        vocab_size (int): Size of vocabulary.
        d_model (int): Hidden dimension of model.
        d_ff (int): Feed-forward expansion dimension.
        h (int): Number of attention heads.
        dropout (float): Dropout probability.
        N (int): Number of Transformer layers.

    Returns:
        torch.Tensor: Output logits shape (B, T, vocab_size).
    """
```

---

## 6. Potential Solution (optional)

**Contributor Tasks:**

* [ ] Add docstrings to all classes in `model/`
* [ ] Add docstrings to classes inside `utils/`
* [ ] Add docstrings to `DataLoader` and training helpers
* [ ] Maintain consistent formatting (NumPy-style or Google-style docstrings)
* [ ] Ensure arguments, returns, and shapes are clearly documented

---

Thank you for taking the time to report this issue.
Your contribution helps improve BardGPT for everyone.
